From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mja00 <me@mja00.dev>
Date: Mon, 12 Aug 2024 06:27:15 -0400
Subject: [PATCH] Add skipping world symlink scan

In worlds that are extremely large (greater than 1TB), it can take an insanely long time to walk the entire world for symlinks.
This patch adds a system property to disable the symlink scan, which can be used to speed up world loading.
A good example of this issue can be found in the Discord (https://discord.com/channels/289587909051416579/289587909051416579/1271527497674129428)
This caused an 18 minute load time for the world. 


diff --git a/src/main/java/net/minecraft/world/level/storage/LevelStorageSource.java b/src/main/java/net/minecraft/world/level/storage/LevelStorageSource.java
index 427ee4d6f12a7abd8da0c65e0b9081b25824df40..42507527dae5156c9d9d93c111571b37fb232dab 100644
--- a/src/main/java/net/minecraft/world/level/storage/LevelStorageSource.java
+++ b/src/main/java/net/minecraft/world/level/storage/LevelStorageSource.java
@@ -420,7 +420,7 @@ public class LevelStorageSource {
 
     public LevelStorageSource.LevelStorageAccess validateAndCreateAccess(String s, ResourceKey<LevelStem> dimensionType) throws IOException, ContentValidationException { // CraftBukkit
         Path path = this.getLevelPath(s);
-        List<ForbiddenSymlinkInfo> list = this.worldDirValidator.validateDirectory(path, true);
+        List<ForbiddenSymlinkInfo> list = Boolean.getBoolean("Paper.disableSymlinkValidation") ? List.of() : this.worldDirValidator.validateDirectory(path, true); // Paper - add skipping of symlinks scan
 
         if (!list.isEmpty()) {
             throw new ContentValidationException(path, list);
